GooglePlayInstant/Editor/QuickDeploy/GcpClient.cs

```diff
@@ -0,0 +1,324 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using UnityEngine;
+
+namespace GooglePlayInstant.Editor.QuickDeploy
+{
+    /// <summary>
+    /// Provides methods for interacting with Google Cloud Platform (GCP), e.g. to upload an Asset Bundle file to
+    /// Google Cloud Storage.
+    /// </summary>
+    public static class GcpClient
+    {
+        private const string TokenUpdateFailedExceptionMessage =
+            "Failed to retrieve access token to use for HTTP request";
+
+        /// <summary>
+        /// Executes all the steps required for deploying a file to GCP according to developer's configuration.
+        /// First verifies if configured bucket exists, and creates the bucket if it does not exist. It then uploads
+        /// the file to GCP and sets the visibility of the file to public.
+        /// </summary>
+        public static void DeployConfiguredFile()
+        {
+            VerifyBucketExistence(
+                // To be executed if bucket exists.
+                bucketExistsResponse => { UploadBundleAndMakeItPublic(); },
+                // To be executed if bucket does not exist.
+                bucketNotFoundResponse =>
+                {
+                    CreateBucket(bucketCreationResponse =>
+                    {
+                        var error = bucketCreationResponse.error;
+                        if (!string.IsNullOrEmpty(error))
+                        {
+                            throw new Exception(string.Format("Got error attempting to create bucket: {0}\n{1}", error,
+                                bucketCreationResponse.text));
+                        }
+
+                        Debug.Log("Google Cloud Storage bucket was successfully created.");
+                        UploadBundleAndMakeItPublic();
+                    });
+                });
+        }
+
+        /// <summary>
+        /// Uploads configured file to GCP and makes the file public.
+        /// </summary>
+        /// <exception cref="Exception">Exception thrown if there was an error uploading the bundle or setting the
+        /// visibility of file to public.</exception>
+        private static void UploadBundleAndMakeItPublic()
```

Comment:
Here's a potential way of centralizing the token refresh policy/logic and generally simplifying the code:

```
        private static void InvokeAccessRestrictedAction(
            Action<Action<WWW>> initialAction, Action<WWW> postResponseAction)
        {
            var token = AccessTokenGetter.AccessToken;
            if (token == null)
            {
                AccessTokenGetter.UpdateAccessToken(() => initialAction(postResponseAction));
            }
            else
            {
                initialAction(postResponseAction);
            }
        }

        private static void UploadBundleAndMakePublic()
        {
            InvokeAccessRestrictedAction(UploadBundleFile, uploadBundleWww =>
            {
                if (!string.IsNullOrEmpty(uploadBundleWww.error))
                {
                    throw new Exception(string.Format("Error uploading bundle: {0}\n{1}", uploadBundleWww.error,
                        uploadBundleWww.text));
                }

                Debug.Log("Uploaded bundle to Google Cloud Platform.");
                var response = JsonUtility.FromJson<FileUploadResponse>(uploadBundleWww.text);
                var bucketName = response.bucket;
                var fileName = response.name;
                if (bucketName != QuickDeployConfig.Config.cloudStorageBucketName)
                {
                    throw new Exception(string.Format(
                        "Response bucket name \"{0}\" doesn't match expected name \"{1}\"",
                        bucketName, QuickDeployConfig.Config.cloudStorageBucketName));
                }

                if (fileName != QuickDeployConfig.Config.cloudStorageFileName)
                {
                    throw new Exception("FILL THIS IN AS ABOVE");
                }

                QuickDeployConfig.Config.assetBundleUrl =
                    string.Format("https://storage.googleapis.com/{0}/{1}", bucketName, fileName);

                InvokeAccessRestrictedAction(MakeBundlePublic, makeBundlePublicWww =>
                {
                    var error = makeBundlePublicWww.error;
                    if (!string.IsNullOrEmpty(error))
                    {
                        throw new Exception(string.Format("Error making file public: {0}\n{1}", error,
                            makeBundlePublicWww.text));
                    }

                    Debug.Log("Set visibility of file to public.");
                });
            });
        }

        private static void UploadBundleFile(Action<WWW> postResponseAction)
        {
            // See https://cloud.google.com/storage/docs/uploading-objects
            var uploadEndpoint =
                string.Format("https://www.googleapis.com/upload/storage/v1/b/{0}/o?uploadType=media&name={1}",
                    QuickDeployConfig.Config.cloudStorageBucketName, QuickDeployConfig.Config.cloudStorageFileName);
            var assetBundleFileBytes = File.ReadAllBytes(QuickDeployConfig.Config.assetBundleFileName);
            var request =
                SendAuthenticatedPostRequest(
                    uploadEndpoint,
                    assetBundleFileBytes,
                    "application/octet-stream",
                    AccessTokenGetter.AccessToken.access_token);
            WwwRequestInProgress.TrackProgress(request, "Uploading file to Google Cloud Storage", postResponseAction);
        }

        private static void CreateBucket(Action<WWW> onCreateBucketResponseAction)
        {
            var credentials = OAuth2Credentials.GetCredentials();
            // see https://cloud.google.com/storage/docs/creating-buckets on creating buckets.
            var createBucketEndPoint = string.Format("https://www.googleapis.com/storage/v1/b?project={0}",
                credentials.project_id);
            var createBucketRequest = new CreateBucketRequest
            {
                name = QuickDeployConfig.Config.cloudStorageBucketName
            };

            var jsonBytes = Encoding.UTF8.GetBytes(JsonUtility.ToJson(createBucketRequest));
            var createBucketWww =
                SendAuthenticatedPostRequest(
                    createBucketEndPoint,
                    jsonBytes,
                    "application/json",
                    AccessTokenGetter.AccessToken.access_token);
            WwwRequestInProgress.TrackProgress(createBucketWww,
                string.Format("Creating bucket with name \"{0}\"", createBucketRequest.name),
                onCreateBucketResponseAction);
        }
```

Reply:
Thanks for the neat approach. Changes all methods sending access restricted requests to follow the same flow.

And about infinite recursion if the token fails to be updated, I AccessTokenGetter.UpdateAccessToken promise that it will throw or invoke a method that throws an error when the requesting the token fails.